Mathlib/RingTheory/LocalRing/Module.lean

```diff
@@ -0,0 +1,264 @@
+  `l` is a split injection iff `k ⊗ l` is a (split) injection.
+-/
+
+
+
+variable {R S} [CommRing R] [CommRing S] [Algebra R S]
+
+section
+
+variable [LocalRing R] {M N} [AddCommGroup M] [AddCommGroup N] [Module R M] [Module R N]
+
+open Function (Injective Surjective Exact)
+open LocalRing TensorProduct
+
+local notation "k" => ResidueField R
+local notation "𝔪" => maximalIdeal R
+
+variable {P} [AddCommGroup P] [Module R P] (f : M →ₗ[R] N) (g : N →ₗ[R] P)
+variable (hg : Surjective g) (h : Exact f g)
+
+theorem LocalRing.map_mkQ_eq_top {N : Submodule R M} [Module.Finite R M] :
+    N.map (Submodule.mkQ (𝔪 • ⊤)) = ⊤ ↔ N = ⊤ := by
+  constructor
+  · intro hN
+    have : 𝔪 • ⊤ ⊔ N = ⊤ := by simpa using Submodule.comap_mono (f := Submodule.mkQ (𝔪 • ⊤)) hN.ge
+    rw [sup_comm] at this
+    exact top_le_iff.mp (Submodule.le_of_le_smul_of_le_jacobson_bot Module.Finite.out
+      (by rw [jacobson_eq_maximalIdeal]; exact bot_ne_top) this.ge)
+  · rintro rfl; simp
+
+theorem LocalRing.map_mkQ_eq {N₁ N₂ : Submodule R M} (h : N₁ ≤ N₂) (h' : N₂.FG) :
+    N₁.map (Submodule.mkQ (𝔪 • N₂)) = N₂.map (Submodule.mkQ (𝔪 • N₂)) ↔ N₁ = N₂ := by
+  constructor
+  · intro hN
+    have : N₂ ≤ 𝔪 • N₂ ⊔ N₁ := by
+      simpa using Submodule.comap_mono (f := Submodule.mkQ (𝔪 • N₂)) hN.ge
+    rw [sup_comm] at this
+    exact h.antisymm (Submodule.le_of_le_smul_of_le_jacobson_bot h'
+      (by rw [jacobson_eq_maximalIdeal]; exact bot_ne_top) this)
+  · rintro rfl; simp
+
+theorem LocalRing.map_mk_eq_top {N : Submodule R M} [Module.Finite R M] :
+    N.map (TensorProduct.mk R k M 1) = ⊤ ↔ N = ⊤ := by
+  constructor
+  · intro hN
+    letI : Module k (M ⧸ (𝔪 • ⊤ : Submodule R M)) :=
+      inferInstanceAs (Module (R ⧸ 𝔪) (M ⧸ 𝔪 • (⊤ : Submodule R M)))
+    letI : IsScalarTower R k (M ⧸ (𝔪 • ⊤ : Submodule R M)) :=
+      inferInstanceAs (IsScalarTower R (R ⧸ 𝔪) (M ⧸ 𝔪 • (⊤ : Submodule R M)))
+    let f := AlgebraTensorModule.lift (((LinearMap.ringLmapEquivSelf k k _).symm
+      (Submodule.mkQ (𝔪 • ⊤ : Submodule R M))).restrictScalars R)
+    have : f.comp (TensorProduct.mk R k M 1) = Submodule.mkQ (𝔪 • ⊤) := by ext; simp [f]
+    have hf : Function.Surjective f := by
+      intro x; obtain ⟨x, rfl⟩ := Submodule.mkQ_surjective _ x;
+      rw [← this, LinearMap.comp_apply]; exact ⟨_, rfl⟩
+    apply_fun Submodule.map f at hN
+    rwa [← Submodule.map_comp, this, Submodule.map_top, LinearMap.range_eq_top.mpr hf,
+      LocalRing.map_mkQ_eq_top] at hN
+  · rintro rfl; rw [Submodule.map_top, LinearMap.range_eq_top]
+    exact TensorProduct.mk_surjective R M k Ideal.Quotient.mk_surjective
+
+theorem LocalRing.subsingleton_tensorProduct [Module.Finite R M] :
+    Subsingleton (k ⊗[R] M) ↔ Subsingleton M := by
+  rw [← Submodule.subsingleton_iff R, ← subsingleton_iff_bot_eq_top,
+    ← Submodule.subsingleton_iff R, ← subsingleton_iff_bot_eq_top,
+    ← LocalRing.map_mk_eq_top (M := M), Submodule.map_bot]
+
+theorem LocalRing.span_eq_top_of_tmul_eq_basis [Module.Finite R M] {ι}
```

Comment:
Some reference to the tensor product in the name?

Reply:
I think `tmul` means tensor product-multiplication in mathlib?